Dark theme

hedyhli opened the pull request:
<details>
<summary> Old PR description </summary>
I have pulled in @GDWR's work and new changes from main.

I have also went ahead and fixed all the conflicts.

This is intended to supplement the original PR #657, in case anyone working on #362 wishes to base the work on top of this one.

~~Unfortunately I did the pull from main in 2 stages, which created two commits of the same message. I couldn't find a way to rectify this without having to fix the conflicts all over again, which is rather laborious.~~

**NOTE**
As of writing, no significant changes were made to GDWR's original implementation. Only trivial ones which were necessary for fixing conflicts.

**PR Status**
For anyone looking to work on top of it, these things have already been done:
- Live theme toggle
- Set of colors in `settings.py` for dark theme
- Most of the colors throughout the site

The rest of the work, IMO, would involve mostly trivial color fixes and testing for the toggle button.
However, some decisions are to be made about the colors, such as the bright blurple, the (hardcoded-color-heavy) timeline page, and the footers.

</details>

---

This PR attempts to revive the work of previous contributors adding a dark theme for the pydis site. Hopefully, if this does not become another stale PR like the unfortunate ones before it, it should close #362.

## Note to reviewers

 Most of the files in the diff are just about adding bulma color classes so they can be displayed correctly in dark mode, you are encouraged to make use of the `Viewed` checkbox. The bulk of the work in this PR is in these files:

1. `pydis_site/settings.py`
1. `pydis_site/static/js/base/themes.js`
1. `pydis_site/templates/base/navbar.html`
1. `pydis_site/static/css/base/themes.css`
1. `pydis_site/static/css/home/index.css`
1. `pydis_site/static/css/home/index.html`

See, not so many files at all :)

I've made a number of notes throughout the diff to hopefully accompany reading of those changes, in places where changes might be questionable.

## Implementation

### JavaScript

Handles:
- Switch toggle click - apply theme and save preference
- Applying correct theme on page load

<img width="900" alt="excalidraw export" src="https://github.com/python-discord/site/assets/50042066/ed024a78-68ab-4a1b-be75-e63f15a9c977">

[View on excalidraw](https://excalidraw.com/#json=qjE0H2NhmbpHBsosmPwuA,0mTLQehFoPxhOrYm4T5V3Q)

### Themes

- The task of including the actual dark theme is delegated to django-simple-bulma, using a set of colors specific for dark mode, defined in `settings.py`.
- Throughout the site, bulma's classes to control colors are applied so that they can switch accordingly to the site's theme; these are the bulk of this PR's file diffs.
- django-simple-bulma template tag is used and both the default and the dark theme stylesheet is included on every page.
- When applying themes, ~~the stylesheet that is unused for that theme is set to have attribute disabled, so the browser uses only one stylesheet~~ (this old method causes FOUC doing theme switches. Now I'm using only one stylesheet and switching its href attribute instead).

### Changes to GDWR's implementation

- Local storage is used rather than cookies to store the user preference for reasons described here 058b5623.
- Removed JQuery dependency on _every page that didn't used to use JQuery_ by rewriting some of the code that depended on JQuery.
- Some other refactors to the javascript.
- Applied bulma classes to some other parts of the sites to ensure themes display correctly.
- Added setting `data-theme="dark"` in the root html element to make CSS styling easier which targets specifically for dark mode.
- Different way to switch stylesheets which happens to eliminate FOUC

### Known limitations

- UI
  - Existing hardcoded colors throughout the site appears to be arbitrarily chosen to make it look best on the specific case, so I did the same where applicable for the dark colors (mostly drop shadows and border colors)
- UX
  - ~~There is a latency in applying the current theme on page load, because this operation is only done after the page is loaded.~~ (this is now fixed thanks to @wookie184)
  - ~~Sometimes, not always, there is a flash of unstyled content (FOUC) effect when switching themes.~~ This is now fixed with the tradeoff of slightly more brittle javascript code.
  -  FOUC when switching system preference while the site is live. See [this comment](https://github.com/python-discord/site/pull/1176#pullrequestreview-1834627048).
- Code
  - Bulma supports neither CSS variables (see [jgthms/bulma#1837], and [[this branch](https://github.com/jgthms/bulma/tree/bulma-v1-css-variables)]) nor any good theming API to make this easier (the `scheme*` variables, due to limitation of django-simple-bulma requiring all variables to be predefined as bulma SCSS aren't imported first, makes defining them in `settings.py` verbose, even so, using these `scheme` variables leads to inconsistency due to limited support for the full range of bulma color variables (only `black*` and `white*` colors are supported, no `grey*`)).

    What I ended up doing is copying bulma CSS attributes where applicable and modifying only the colors.

## Demo

<details><summary>Live system color scheme preference change</summary>

https://github.com/python-discord/site/assets/50042066/867c2ee7-9f49-43f6-88a5-dfeb60f6b104

</details>

## Thanks

This PR's work stands on the shoulder of giants. Notably, with thanks to:

- Lemon for the beautiful switch design and implementation ideas in the original issue,
- GDWR for initial implementation and bulk of the work on applying the bulma classes,
- Contributors to django-simple-bulma adding the theme feature.

## Todo

- [x] Fix homepage wave
- [x] Fix homepage hardcoded colors
- [x] Fix the hardcoded colors on the timeline page
- [x] Fix colors on contributing page
- [x] Fix colors on resources page
- [x] Look for hardcoded colors elsewhere
- [x] Possibly reconsider how the dark theme is applied so that it is the most minimal, performant, and maintainable (like using [localstorage](https://caniuse.com/?search=localstorage) [rather than cookies](https://web.dev/patterns/theming/theme-switch#js)?)
- [x] Respect `@media` query if user preference not already recorded

--- WIP ➡️  ready for review ---

- [x] Device & browser testing for the switch

GDWR committed:
Correct typehints

GDWR committed:
Add dark theme stylesheet

GDWR committed:
Update dependency

GDWR committed:
Add initial dark theme

GDWR committed:
Resize switch

GDWR committed:
getCurrentTheme returns "default" if empty

GDWR committed:
Comment bulma variables

GDWR committed:
Use colour helpers

GDWR committed:
Use colour helpers

GDWR committed:
Use colour helpers

GDWR committed:
Use colour helpers

GDWR committed:
Set dark theme colour values

GDWR committed:
Use colour helpers

GDWR committed:
Use colour helpers

GDWR committed:
remove unused colour values

GDWR committed:
Determine current theme and set switch

GDWR committed:
Use colour helpers

GDWR committed:
Change wave colour temp

GDWR committed:
Remove clashing css

GDWR committed:
Update colour variables

GDWR committed:
Resize theme switch

GDWR committed:
Resize theme switch

Commit:
Fix all conflicts

hopefully I dont have to do this again

Commit:
Automatically fix lint

Commit:
Dark: Fix most colors on timeline

A LOT of colors are still hardcoded in the CSS

Commit:
Bump django-simple-bulma to 2.6.0

And use `include_js` flag to the bulma dark theme stylesheet template
tag.

Commit:
Shouldn't a white wave be white?

GDWR commented on pydis_site/static/images/waves/wave_white.svg:
Oops lol, thats a hilarious commit

Commit:
Dark: Refactor toggle handling JS

- Use localstorage. Advantages:
  - One key=value for the entire site, without needing to specify
    `path=/`
  - No need for string splitting to parse the `key=value; ...` data
  - Suggested more widely in tutorials

  Pretty good support: https://caniuse.com/?search=localstorage

- Remove the need for JQuery, use IDs for switch and knob elements.
  - This also makes the code more robust if the page has other switch &
    knob classes!

Commit:
Dark: Support system color scheme preference

Added features
- When no preference already stored when page loads, we'll take into
  account the system setting and set the theme automatically.
- When the system setting is modified while the site is active, switch
  theme automatically.
- The system preference takes lower precedence than saved preference
  from the switch toggle, unless the setting is modified while the site
  is active.

Miscellaneous refactor.

Commit:
Dark: Fix hero colors on homepage

We can now use `[data-theme="dark"]` selector in CSS to defined styles
specifically for dark mode. This follows common web dev patterns and
gives little overhead.

- Hero
  - Added `wave_black` SVG which copies `wave_white` as is but uses the
    site's dark mode background color.

Commit:
Dark: Inline selectors for the hero bottom wave

Commit:
Remove obsolete comment regarding notification.css file

Commit:
Dark: Fix shadows & borders colors on the homepage

hedyhli commented on pydis_site/context_processors.py:
```diff
@@ -1,7 +1,7 @@
 from django.conf import settings
-from django.template import RequestContext
+from django.http import HttpRequest
 
 
-def git_sha_processor(_: RequestContext) -> dict:
+def git_sha_processor(_: HttpRequest) -> dict:
```
Hi @GDWR, it's totally fine if you don't remember anymore, but would you happen to know why this change was required in your original PR?

GDWR replied:
its been a while, but iirc the type was incorrect. And just thought it would be nice to correct it incase anybody ever tries to use the param.

hedyhli replied:
Oh I see. Thanks!

hedyhli commented on pydis_site/static/images/waves/wave_black.svg:
```diff
@@ -0,0 +1,77 @@
+     fill-rule="evenodd"
+     d="M 1599.995,17.566918 C 1289,17.566918 1190.102,-0.03623696 789,5.6042811e-5 389,5.6042811e-5 289,17.566918 0,17.566918 v 11.178914 h 1600 c 0,0 -0.01,-6.968673 -0.01,-11.178914 z"
+     id="path9"
+     style="fill:#2C2F33;fill-opacity:1;stroke-width:0.381026" />
```
This file was copied as-is from `wave_white.svg`, only changing the fill color to `#2C2F33`, the primary site background color defined in `settings.py`.

hedyhli commented on pydis_site/static/css/home/index.css:
```diff
@@ -107,6 +111,10 @@ h1 {
 }
 
+[data-theme="dark"] #showcase .box {
+    box-shadow: 0 0.5em 1em -0.125em rgb(10 10 10 / 10%), 0 0px 0 1px rgb(10 10 10 / 2%);
```
This was copied as-is from bulma's own code, only changing the shadow colors to match those of the project cards.

Commit:
Dark: Revert some not so decent whitespace changes from previous PR

hedyhli reviewed (commented):
Leaving some notes for (future) reviewers.

Commit:
Dark: Name JS & CSS files according to their content

We have the CSS for the navbar in base.css, the relevant files only
contain code for the theme toggle, so they should be named as such.

If we ever implement CSS variables per-theme, they could then be done in
`themes.css`.

Commit:
Dark: Fix all link hover colors (hopefully) everywhere

Commit:
Dark: Fix colors for bulma cards shadow & borders

Commit:
Dark: Fix backgrounds on sponsors section on homepage

See https://discord.com/channels/267624335836053506/635950537262759947/1193170173574774855

Idea from go.dev -- add background to each sponsor logo and put them in
a grid.

Also set the sponsor section background color to match that of footer's.

Issue pointed out by @ChrisLovering on discord.

This is an initial draft. As of now this works from my own testing:

- Consistent & uniform spacing (gaps)
- Supports mobile
- No affect to light theme other than the arrangement
- Logos do not get squashed, ever

Commit:
Dark: Fix sizing of sponsors' logos

Also works on mobile.

I *love* CSS.

Commit:
Dark: Fix timeline color for borders

Commit:
Dark: Use link color of better contrast

Commit:
Dark: Fix colors on content pages

Commit:
Merge branch 'main' into gdwr-dark-theme

Commit:
Dark: Fix tag colors in resources page

Commit:
Merge branch 'gdwr-dark-theme' of github.com:hedyhli/pydis-site into gdwr-dark-theme

Commit:
Dark: Fix the filter box on resources page

Commit:
Dark: Use adafruit logo with light background in resources

Commit:
Dark: Use kivy logo with light background in resources

Commit:
Fix typo in page.css

hedyhli commented on pydis_site/static/images/resources/kivy.png:
Elements in this image were placed in such a way so it looks the same as before (in light theme before this PR).

Commit:
Remove extra trailing space in 8th codejam page

Commit:
Remove extra trailing space in resources.html

hedyhli commented on pydis_site/apps/resources/resources/automate_the_boring_stuff_book.yaml:
I changed all "black" icons in resources to use dark so they can show up correctly in dark theme. For some reason I couldn't make black icons show as white without writing the CSS manually.

The tradeoff is that they're all blurple now due to being links.

hedyhli commented on pydis_site/apps/resources/resources/kivy.yaml:
The existing kivy title image is dark and transparent making it hard to read in dark mode.

I've merged the icon and title image together into one image so it appears more consistent with other title images.

hedyhli commented on pydis_site/settings.py:
These colors were adjusted with accessibility in mind.

django-simple-bulma does not support including the default variables into per-theme variables, so those non-color related variables have to be repeated for dark theme.

It also doesn't support reusing of default Bulma variables, so we have to define it ourselves before using it (in which case I've copied Bulma's colors for it).

hedyhli commented on pydis_site/static/css/home/index.css:
```diff
@@ -218,21 +241,72 @@ h1 {
-    margin: auto;
-    max-width: 80%;
+[data-theme="dark"] #sponsors {
+    background-color: #2C2F33 !important;
```
The sponsors section have been revamped so they can be displayed nicely readable in dark mode. This also happens to make their sizes more consistent for light mode (see #662).

Also see [relevant discord discussion](https://discord.com/channels/267624335836053506/635950537262759947/1193170173574774855).

hedyhli commented on pydis_site/templates/home/index.html:
```diff
@@ -181,35 +181,39 @@ <h1 class="is-size-1">Projects</h1>
         <h1 class="title is-6 has-text-grey">
           Sponsors
         </h1>
-        <div class="columns is-mobile is-multiline">
```
Read together with [this](https://github.com/python-discord/site/pull/1176#discussion_r1454365099) comment.

hedyhli reviewed (commented):
Some more notes.

Commit:
Merge branch 'main' into gdwr-dark-theme

Commit:
Consistent 4-spaces CSS indent

Commit:
Dark: Sponsors light background for code/gamejams

Commit:
Fix conflicts

Commit:
Merge branch 'main' into gdwr-dark-theme

Commit:
Consistent CSS indent for files that I touched

Commit:
Dark: For content pages using transparent images...

...which are unreadable in dark mode, use `has-dark-mode-background`
class.

    ![](/img.png){: class="has-dark-mode-background" }

Like this. And it will be given a nice light gray background and a
subtle padding & border.

Commit:
Dark: Improve order of switching stylesheets

wookie184 reviewed (changes requested):
This is really good! This review is just from me playing around with the site in firefox, on desktop. I'll try an have a check in chrome and on mobile when I get a chance, and have a better look at the code changes, but honestly i'm not sure i'll have much else to say since this seems great to me already!

-----
Regarding
> There is a latency in applying the current theme on page load, because this operation is only done after the page is loaded.

which causes a small flash on the page in some cases, I think we can safely set the style sheets before the page is loaded by moving this code out of the `DOMContentLoaded` listener which should prevent flashing.
```js
let theme;
const systemPrefersDark = window.matchMedia("(prefers-color-scheme: dark)");

if (systemPrefersDark.matches)
    theme = getCurrentTheme("dark");
else
    theme = getCurrentTheme();

setStyleSheets(theme);
```
As far as my HTML/browser understanding goes, this should work fine as long as the script tag is below the stylesheet tags it modifies, which it is currently.

-----
The titles for projects in light mode are now lighter than they were before. I think this makes them harder to read as the contrast is lower.

Before:
![image](https://github.com/python-discord/site/assets/22353562/27b9c60f-5ee6-4688-994b-99f1c65575dc)

After:
![image](https://github.com/python-discord/site/assets/22353562/cc6b1a41-693e-401f-b51e-8096ac36dfd6)

----

The [panda3d logo on the resources page](https://deploy-preview-1176--pydis-static.netlify.app/resources/?search=panda3d) doesn't display on dark mode very well:

![image](https://github.com/python-discord/site/assets/22353562/b9837e5e-8594-4960-bffb-03557e24f7b3)

We could use a non transparent image, or only use the panda face icon and put the "panda3d" part as the title text. All the other ones seem good.

Commit:
Dark: Set correct theme immediately when the script loads

This fixes the latency previously observed on page loads.

We still have to put switch toggling after page loads though.

Genius idea by wookie.

Co-authored-by: wookie184 <wookie1840@gmail.com>

Commit:
Dark: Possibly fix FOUC during theme switch

Commit:
Dark: Use original light theme colors everywhere

Adjusted box-shadow for dark mode to be more readable, unfortunately
this gave it a little neomorphism effect...

Commit:
Temporary fix to theme-switching FOUC problem

Commit:
Dark: Use our own panda3d logo with light background

Commit:
Dark: More subtle collapsible and box-shadow borders

Commit:
Optimistic fix for FOUC on initial site load

Commit:
Dark: Theme switch with better contrast on mobile

Commit:
Dark: Avoid knob & icon overlapping during transition

Commit:
Dark: Better contrast for dark switch

Commit:
Dark: Better shadows

Commit:
Dark: Restore border on github cards

wookie184 commented on pydis_site/static/js/base/themes.js:
```diff
@@ -0,0 +1,92 @@
+
+    const switchToggle = document.getElementById("theme-switch");
+    switchToggle.addEventListener("click", () => {
+        const newTheme = getCurrentTheme() === "light" ? "dark" : "light";
```
If you have nothing set in localstorage and are on a system that prefers dark mode, the site will display in dark mode, but you have to click twice to change to light mode.

I believe this is because `getCurrentTheme` will return `light` here if there is nothing in local storage even if the current theme is dark mode because of system preference. Passing the `systemPrefers` argument here should fix it.

hedyhli replied:
This is a good catch.

Your suggested solution does fix it. However after implementing it I decided that it made the logic and integration with `getCurrentTheme` it little complicated.

In the switch toggle our intention is to toggle the current theme. Since the `data-theme` attribute is set for all of system preference changes, switch clicks, and page loads, why not check for that instead? It makes the code cleaner and more explicit IMO. And it cannot, ever, contradict what the user actually sees the theme to be on the site.

The localStorage would then only be used when the page is first loaded to restore preference from last time.
The system preference would serve only two purposes: obtain correct theme if no localStorage preference stored, apply new theme if the system preference changes.

That's just my opinion. I can force-push any alternative implementations if you prefer.

Thanks for your thorough testing!

wookie184 replied:
Yeah, this is nicer, good stuff 👌

wookie184 reviewed (changes requested):
Tested this out on firefox/chrome on desktop and on firefox on my phone and all seems good. Just one minor thing I noticed then this should be good as far as i'm concerned.

Commit:
Dark: Fix switch toggle when system preference is used

Thanks to wookie for finding the bug.

hedyhli reviewed (commented):
By the way, I am observing flash of unstyled content when I change the system preference and have the site automatically adapt to the new theme. It might have been caused by me fixing FOUC for switch toggle clicks, and/or FOUC on page loads.

I will probably not be fixing this for two reasons:
1. No idea how to fix for now
2. This feature is used much less often than the switch toggle, I only implemented it for completeness

If anyone has a fix though, I would gladly apply it.

wookie184 reviewed (approved):
Perfect 🚀

lemonsaurus reviewed (approved):
Hi @hedyhli. This PR is absolutely _beautiful_ work, and it's maybe one of the best PR descriptions I've ever seen in my entire career of managing software developers. Bravo. 👏🏼 👏🏼 

The code is well documented, well implemented, and you've been extremely thorough. I have notes, but they are all cosmetic notes and everything in this PR is more or less perfectly implemented already.

# The colors
My main complaint is simply that I'm not a big fan of the colors. 

### Darker background
I think the background should be darker, and there should be more contrast between the background and the cards, for example, this is with the background set to `#1e2329` - note the contrast now between background and cards, I quite like it.
![image](https://github.com/python-discord/site/assets/2098517/2a79efe6-10b8-44d4-9e13-4bdbc3def1b1)

### Mismatched switcher purple
I don't really think this purple works well with the navbar purple:
![image](https://github.com/python-discord/site/assets/2098517/03e98141-8a9c-4b86-8ea2-fc61051d054c)

### Navbar is too damn bright 😎 
Perhaps most of all, and this may be controversial, I think the navbar is far too bright when dark mode is enabled. I'd propose that we change the navbar color to be darker. For example, here it is using `#3b4774`:
![image](https://github.com/python-discord/site/assets/2098517/7a310725-a732-48b0-9521-9b28453d0e6b)

### Darker discord component
I also think the discord component in the top right should change along with it. This would require altering the underlaying svg to maybe the same as the html background, and then changing the background-color of the <a> component to something like `#485580`:
![image](https://github.com/python-discord/site/assets/2098517/b410ab72-78ed-4a5b-8b72-a25617beb165)

# Other small stuff
I also have a few minor things that annoy me, these are not very important but I'll list them anyway.

### Sponsor logos
I get why we did this, but having a bunch of bright white boxes on the front page during dark mode irks me. Some of these logos do have dark-mode friendly alternatives available in their press kits and branding pages. For the others, I would suggest we just make some. Happy to help with this.
![image](https://github.com/python-discord/site/assets/2098517/1b730cc9-c1b4-4a4e-a110-6a52afda8265)

### Resource logos
Many of the resources have icky dark theme logos as well, this would also be great to fix by just creating a set of dark theme logos for each resource. 
![image](https://github.com/python-discord/site/assets/2098517/257e0245-88df-4849-841c-cf6b800866f7)

### Timeline icons
Not a huge fan of the timeline icons having white backgrounds, as shown here:
![image](https://github.com/python-discord/site/assets/2098517/6409d558-909d-46e6-8a09-8084f785329d)

Here's an example where the icon is darkened a bit and the circle around it is the same color as the line between the icons:
![image](https://github.com/python-discord/site/assets/2098517/94bd38ac-96d2-4b56-8fde-ca032b984078)

### Various buttons
There are a few buttons that also use the very bright navbar color, like the `"Check it out!"` button on the main page in the timeline widget. Here's an example where that one is darker:
![image](https://github.com/python-discord/site/assets/2098517/e78d70f9-ea93-48b9-9633-05c0cc73adad)

# But should we merge this?
I mean, **maybe**. It's functional and these changes could all be in a separate PR, and would be easier to review and discuss if they were. It might be better for us to just merge this, migrate this review into a new issue, and tackle those changes separately. I certainly don't want to be the reason this PR goes stale.

I think you've done some tremendous work here, @hedyhli, and I'd love to see it in production very soon. On the other hand, if you are looking at these suggestions and thinking "oh okay that's easy, most of this will take me 10 minutes" and you agree with all my feedback, feel free to sneak a few of them in before we merge. It's your call. I'm approving this and you can merge if you'd like.

Commit:
Merge branch 'main' into gdwr-dark-theme

Commit:
Dark: Darken nav, hero, hero waves, button, body backgrounds

Commit:
Dark: Fix timeline icon borders

Commit:
Dark: Use alternate images for homepage, resources, events

Thanks to lemon for editting all the images.

Co-authored-by: Leon Sandøy <leon.haland@gmail.com>

Xithrius reviewed (approved):
Approval for good luck

:shipit:

Commit:
Dark: Use a darker pydis blurple everywhere
